### PR TITLE
[Feature] Support SemanticKitti Dateset

### PR DESCRIPTION
**Pull request by Jacky-gsq**
This PR is for supporting the semantic-Kitti dataset and is related to the issue #1417 #1350

It mainly includes:

1. The implementation of `evaluate` function in` SemanticKittiDataset`
2. The pipeline for loading the `calib` and `pose` information in the dataset
3. The unit test for the above code

There are still some problems with details and looking forward to your advice.
@Tai-Wang @ZwwWayne Sorry, I'm so stupid and can't seem to find the button of “adding reviewer”. So I mentioned you.

### CODE REVIEW

File: configs/_base_/datasets/semantic_kitti.py [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+        shift_height=True,
+        load_dim=4,
+        use_dim=[0, 1, 2]),
+    dict(
+        type='LoadAnnotations3D',
+        with_bbox_3d=False,
+        with_label_3d=False,
+        with_mask_3d=False,
+        with_seg_3d=True,
+        seg_3d_dtype=np.int32),
```
Comment (Tai-Wang) [inline configs/_base_/datasets/semantic_kitti.py:46]: This type can be set by default when loading annotations. It is a little inappropriate to set it here.

File: mmdet3d/datasets/pipelines/loading.py [hunk 2]
```diff
@@ -238,6 +240,104 @@ def __repr__(self):
         return f'{self.__class__.__name__}(sweeps_num={self.sweeps_num})'
 
 
+@PIPELINES.register_module()
+class LoadCalibPoseTime(object):
+    """load calib poses and times for SemanticKitti dataset."""
```
Comment (Tai-Wang) [inline mmdet3d/datasets/pipelines/loading.py:245]: Please complement the docstring following the format of other pipelines.

File: mmdet3d/datasets/pipelines/loading.py [hunk 3]
```diff
@@ -238,6 +240,104 @@ def __repr__(self):
...
+
+    def __init__(self, use_calib=True, use_pose=True, use_time=False):
+
+        # self.data_root = data_root
```
Comment (Tai-Wang) [inline mmdet3d/datasets/pipelines/loading.py:249]: Please remove unused comments.

File: mmdet3d/datasets/pipelines/loading.py [hunk 4]
```diff
@@ -238,6 +240,104 @@ def __repr__(self):
...
+        self.use_calib = use_calib
+        self.use_pose = use_pose
+        self.use_time = use_time
+
+    def __call__(self, results):
+
+        seq_folder = results['ann_info']['pts_semantic_mask_path'].replace(
+            'labels/000000.label', '')
```
Comment (Tai-Wang) [inline mmdet3d/datasets/pipelines/loading.py:257]: It's a hard code here. And why only replace 000000.label?

File: mmdet3d/datasets/pipelines/loading.py [hunk 5]
```diff
@@ -238,6 +240,104 @@ def __repr__(self):
...
+            # Read poses
+            poses_f64 = self.parse_poses(
+                osp.join(seq_folder, 'poses.txt'), calibrations)
+            results['poses'] = [pose.astype(np.float32) for pose in poses_f64]
+
+        return results
+
+    def parse_calibration(self, filename):
+        """ read calibration file with given filename
+            Returns
```
Comment (Tai-Wang) [inline mmdet3d/datasets/pipelines/loading.py:280]: Please follow the format of other functions to refine the docstring

File: mmdet3d/datasets/pipelines/loading.py [hunk 6]
```diff
@@ -600,11 +700,11 @@ def _load_masks_3d(self, results):
             self.file_client = mmcv.FileClient(**self.file_client_args)
         try:
             mask_bytes = self.file_client.get(pts_instance_mask_path)
-            pts_instance_mask = np.frombuffer(mask_bytes, dtype=np.int64)
+            pts_instance_mask = np.frombuffer(mask_bytes, dtype=np.int)
```
Comment (Tai-Wang) [inline mmdet3d/datasets/pipelines/loading.py:703]: Why do here use np.int while others use np.long?

File: mmdet3d/datasets/pipelines/loading.py [hunk 7]
```diff
@@ -518,7 +618,7 @@ def __init__(self,
                  with_seg=False,
                  with_bbox_depth=False,
                  poly2mask=True,
-                 seg_3d_dtype=np.int64,
+                 seg_3d_dtype='int',
```
Comment (Tai-Wang) [inline mmdet3d/datasets/pipelines/loading.py:621]: Why do we need to modify this default setting?

File: mmdet3d/datasets/semantickitti_dataset.py [hunk 8]
```diff
@@ -60,6 +67,32 @@ def __init__(self,
...
+
+        self.include = np.array(
+            [n for n in range(self.n_classes) if n not in self.ignore],
+            dtype=np.int64)
+        self.offset = offset
+        self.eps = 1e-15
+        self.min_points = min_points
+        self.reset()
+        self.global_cfg = mmcv.load(
+            osp.join(self.data_root, 'semantic-kitti.yaml'))
```
Comment (Tai-Wang) [inline mmdet3d/datasets/semantickitti_dataset.py:81]: where is the yaml file?

File: mmdet3d/datasets/semantickitti_dataset.py [hunk 9]
```diff
@@ -108,3 +141,281 @@ def get_ann_info(self, index):
...
+                gt_annos, result_files)
+            print_log('\n' + ap_result_str, logger=logger)
+
+        if tmp_dir is not None:
+            tmp_dir.cleanup()
+        if show or out_dir:
+            self.show(results, out_dir, show=show, pipeline=pipeline)
+        return ap_dict_list
+
+    def semantic_kitti_eval(self, gt_annos, result_files):
```
Comment (Tai-Wang) [inline mmdet3d/datasets/semantickitti_dataset.py:221]: These specific evaluation codes can be moved to mmdet3d/core/evaluation/semantickitti_eval.py like other datasets.

File: tests/test_data/test_datasets/test_semantickitti_dataset.py [hunk 10]
```diff
@@ -51,3 +62,82 @@ def test_getitem():
...
+    global_cfg = mmcv.load(osp.join(data_root, 'semantic-kitti.yaml'))
+    class_remap = global_cfg['learning_map']
+    inst_pred_0 = np.fromfile(
+        osp.join(data_root, 'sequences/00/labels/000000.label'),
+        dtype=np.int32).reshape(-1, 1)
+    sem_pred_0 = inst_pred_0 & 0xFFFF
+    sem_pred_0 = np.vectorize(class_remap.__getitem__)(sem_pred_0)
+
+    # sem_pred_0, inst_pred_0, sem_gt_0, inst_gt_0 = gen_psuedo_labels(50)
+    # sem_pred_1, inst_pred_1, sem_gt_1, inst_gt_1 = gen_psuedo_labels(51)
```
Comment (Tai-Wang) [inline tests/test_data/test_datasets/test_semantickitti_dataset.py:131]: Please remove unused codes.

File: tests/test_data/test_datasets/test_semantickitti_dataset.py [hunk 11]
```diff
@@ -51,3 +62,82 @@ def test_getitem():
...
+    result0 = dict(pred_sem=sem_pred_0, pred_inst=inst_pred_0)
+
+    ap_dict = semantic_kitti_dataset.evaluate([result0])
+
+    assert np.isclose(ap_dict[0]['all']['PQ'], 0.21052631578947367)
+    assert np.isclose(ap_dict[0]['all']['SQ'], 0.21052631578947367)
+    assert np.isclose(ap_dict[0]['all']['RQ'], 0.21052631578947367)
+
+
+if __name__ == '__main__':
```
Comment (Tai-Wang) [inline tests/test_data/test_datasets/test_semantickitti_dataset.py:141]: Please remove line 141-143

**Comment by Jacky-gsq** [comment]
OK, I will fix the above bugs at the weekends! Thank your reviews.

**Comment by Xiangxu-0103** [comment]
I think it is better to generate info files like general datasets do first. If possible, I will help to support semantickitti semantic segmentation dataset when free.

**Comment by Tai-Wang** [comment]
Please "resolve" the previous comments when you fix them, and then we can have the next round review. Thanks for your great efforts!

File: configs/_base_/datasets/semantic_kitti.py (hunk 1 above)
Comment (Jacky-gsq) [reply configs/_base_/datasets/semantic_kitti.py:46]: That has been deleted in the new version.

Comment (Jacky-gsq) [reply configs/_base_/datasets/semantic_kitti.py]: The setting about "LoadAnnotations3D" has been deleted.

File: mmdet3d/datasets/pipelines/loading.py (hunk 2 above)
Comment (Jacky-gsq) [reply mmdet3d/datasets/pipelines/loading.py]: All the docstrings have been added.

File: mmdet3d/datasets/pipelines/loading.py (hunk 3 above)
Comment (Jacky-gsq) [reply mmdet3d/datasets/pipelines/loading.py]: The unused comments in all files have been removed.

File: mmdet3d/datasets/pipelines/loading.py (hunk 4 above)
Comment (Jacky-gsq) [reply mmdet3d/datasets/pipelines/loading.py]: I replace the `000000.label` by `sample_idx`.

```
sample_idx = results['sample_idx']
seq_folder = results['ann_info']['pts_semantic_mask_path'].replace(f'labels/{sample_idx}.label', '')
```

File: mmdet3d/datasets/pipelines/loading.py (hunk 7 above)
Comment (Jacky-gsq) [reply mmdet3d/datasets/pipelines/loading.py]: Sorry for the fault. I seem to commit the code based on my local version of mmdet3d last time and I have rectified that as the default setting.

File: mmdet3d/datasets/semantickitti_dataset.py (hunk 8 above)
Comment (Jacky-gsq) [reply mmdet3d/datasets/semantickitti_dataset.py]: The related parameters "learning_map" in the yaml file have been added as the global setting in `class SemanticKITTIDataset`

File: mmdet3d/datasets/semantickitti_dataset.py (hunk 9 above)
Comment (Jacky-gsq) [reply mmdet3d/datasets/semantickitti_dataset.py]: I have migrated the evaluation codes to the above path, but by a way of `class SemanticKITTIEval(object)`. Is that ok?

File: mmdet3d/core/evaluation/semantickitti_eval.py [hunk 12]
```diff
@@ -0,0 +1,232 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import numpy as np
+
+
+class SemanticKITTIEval(object):
```
Comment (Tai-Wang) [inline mmdet3d/core/evaluation/semantickitti_eval.py:5]: Does this class come from the official devkit? If not, please add corresponding docstrings.

File: mmdet3d/datasets/pipelines/loading.py [hunk 13]
```diff
@@ -238,6 +240,133 @@ def __repr__(self):
         return f'{self.__class__.__name__}(sweeps_num={self.sweeps_num})'
 
 
+@PIPELINES.register_module()
+class LoadCalibPoseTime(object):
```
Comment (Tai-Wang) [inline mmdet3d/datasets/pipelines/loading.py:244]: This pipeline should be placed in the get_data_info of SemanticKITTI Dataset.

Comment (Tai-Wang) [reply mmdet3d/datasets/pipelines/loading.py:244]: Please refer to other datasets for the way to load calibration/pose/timestamps. (We only load and process the point/img data in the pipeline.)

File: mmdet3d/datasets/semantickitti_dataset.py [hunk 14]
```diff
@@ -40,12 +44,51 @@ class SemanticKITTIDataset(Custom3DDataset):
                'person', 'bicyclist', 'motorcyclist', 'road', 'parking',
                'sidewalk', 'other-ground', 'building', 'fence', 'vegetation',
                'trunck', 'terrian', 'pole', 'traffic-sign')
+    learning_map = {
```
Comment (Tai-Wang) [inline mmdet3d/datasets/semantickitti_dataset.py:47]: label_mapping / class_mapping would be better?

File: mmdet3d/datasets/semantickitti_dataset.py [hunk 15]
```diff
@@ -108,3 +153,82 @@ def get_ann_info(self, index):
 
         anns_results = dict(pts_semantic_mask_path=pts_semantic_mask_path)
         return anns_results
+
+    def parse_label(self, filename):
```
Comment (Tai-Wang) [inline mmdet3d/datasets/semantickitti_dataset.py:157]: Need a brief docstring

File: mmdet3d/models/decode_heads/dsnet_head.py [hunk 16]
```diff
@@ -0,0 +1,118 @@
...
+from .decode_head import Base3DDecodeHead
+
+if IS_SPCONV2_AVAILABLE:
+    from spconv.pytorch import SparseSequential
+else:
+    from mmcv.ops import SparseSequential
+
+
+@HEADS.register_module()
+class DSNetSemHead(Base3DDecodeHead):
```
Comment (Tai-Wang) [inline mmdet3d/models/decode_heads/dsnet_head.py:18]: The DSNet model needs to be moved to another new PR.

File: tests/test_data/test_datasets/test_semantickitti_dataset.py [hunk 17]
```diff
@@ -51,3 +61,110 @@ def test_getitem():
...
+def test_evaluate():
+    if not torch.cuda.is_available():
+        pytest.skip('test requires GPU and torch+cuda')
+    data_root = './tests/data/semantickitti/'
+    ann_file = './tests/data/semantickitti/semantickitti_infos.pkl'
+    classes = ('unlabeled', 'car', 'bicycle', 'motorcycle', 'truck', 'bus',
+               'person', 'bicyclist', 'motorcyclist', 'road', 'parking',
+               'sidewalk', 'other-ground', 'building', 'fence', 'vegetation',
+               'trunck', 'terrian', 'pole', 'traffic-sign')
+    learning_map = {
```
Comment (Tai-Wang) [inline tests/test_data/test_datasets/test_semantickitti_dataset.py:75]: label_mapping / class_mapping

**Comment by ShihaoSu** [comment]
> This PR is for supporting the semantic-Kitti dataset and is related to the issue #1417 #1350
> 
> It mainly includes:
> 
> 1. The implementation of `evaluate` function in` SemanticKittiDataset`
> 2. The pipeline for loading the `calib` and `pose` information in the dataset
> 3. The unit test for the above code
> 
> There are still some problems with details and looking forward to your advice. @Tai-Wang @ZwwWayne Sorry, I'm so stupid and can't seem to find the button of “adding reviewer”. So I mentioned you.

Thx for your work. I would appreciate it if you can show how you format the `anno_file`.

File: tests/test_data/test_datasets/test_semantickitti_dataset.py (hunk 17 above)
Comment (Jacky-gsq) [reply tests/test_data/test_datasets/test_semantickitti_dataset.py:75]: OK

File: mmdet3d/datasets/semantickitti_dataset.py [hunk 18]
```diff
@@ -108,3 +153,90 @@ def get_ann_info(self, index):
...
+        """Parse the label files of SemanticKitti dataset.
+
+        Args:
+            filename (string): The label files path.
+
+        Returns:
+            gt_annos:(list [dict]): The semantic and instance labels info.
```
Comment (Tai-Wang) [inline mmdet3d/datasets/semantickitti_dataset.py:164]: list[dict]

Comment (Tai-Wang) [inline mmdet3d/datasets/semantickitti_dataset.py:164]: list[dict]

**Comment by Tai-Wang** [comment]
Please fix the linting. You can refer to the anno_file format of KITTI. It saves the filename, calibration and pose information in the anno_file such that we can directly load them in get_data_info in the dataset class.

**Comment by Tai-Wang** [comment]
Please fix the linting. You can refer to the anno_file format of KITTI. It saves the filename, calibration and pose information in the anno_file such that we can directly load them in get_data_info in the dataset class.

**Comment by Jacky-gsq** [comment]
Get! I will modify the functions to generate the anno_files, referring to the KITTI dataset.

**Pull request closed without merging**